x-pack/metricbeat/module/mssql/transaction_log/transaction_log.go

```diff
@@ -0,0 +1,217 @@
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/x-pack/metricbeat/module/mssql"
+)
+
+type dbInfo struct {
+	id   int
+	name string
+}
+
+type logSpace struct {
+	id                             int
+	totalLogSizeInBytes            int
+	usedLogSpaceInBytes            int
+	usedLogSpaceInPercent          float64
+	logSpaceInBytesSinceLastBackup int
+}
+
+type logStats struct {
+	databaseID            int
+	sizeMB                float64
+	activeSizeMB          float64
+	backupTime            string
+	sinceLastBackupMB     *float64
+	sinceLastCheckpointMB float64
+	recoverySizeMB        float64
+}
+
+func init() {
+	mb.Registry.MustAddMetricSet("mssql", "transaction_log", New,
+		mb.DefaultMetricSet(),
+		mb.WithHostParser(mssql.HostParser))
+}
+
+// MetricSet type defines all fields of the MetricSet
+type MetricSet struct {
+	mb.BaseMetricSet
+	log *logp.Logger
+	db  *sql.DB
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	cfgwarn.Beta("The mssql transaction_log metricset is beta.")
+
+	logger := logp.NewLogger("mssql.transaction_log").With("host", base.HostData().SanitizedURI)
+
+	db, err := mssql.NewConnection(base.HostData().URI)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not create connection to db")
+	}
+
+	return &MetricSet{
+		BaseMetricSet: base,
+		log:           logger,
+		db:            db,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(reporter mb.ReporterV2) {
+	dbs, err := m.getDbsNames()
+	if err != nil {
+		reporter.Error(err)
+		return
+	}
+
+	for _, db := range dbs {
+		moduleFields := common.MapStr{
+			"database": common.MapStr{
+				"id":   db.id,
+				"name": db.name,
+			},
+		}
+		metricsetFields := common.MapStr{}
+
+		spaceUsage, err := m.getLogSpaceUsageForDb(db.name)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["space_usage"] = spaceUsage
+		}
+
+		stats, err := m.getLogStats(db)
+		if err != nil {
+			reporter.Error(err)
+		} else {
+			metricsetFields["stats"] = stats
+		}
+
+		if len(metricsetFields) == 0 {
+			m.log.Debug("no data to report")
+			continue
+		}
+
+		// Both log space and log size are available, so report both
+		if isReported := reporter.Event(mb.Event{
+			ModuleFields:    moduleFields,
+			MetricSetFields: metricsetFields,
+		}); !isReported {
+			m.log.Debug("event not reported")
+		}
+
+	}
+}
+
+// Close the connection to the server at the engine level
+func (m *MetricSet) Close() error {
+	return m.db.Close()
+}
+
+func (m *MetricSet) getLogSpaceUsageForDb(dbName string) (result common.MapStr, err error) {
```

Comment:
Not sure of the benefits of using named return values here, but ok.

Reply:
I personally prefer not to use named returns and only if it is really need like in defer functions. It makes from my perspective code more readable. @sayden could you update this?

Reply:
Changed. I see the benefits and I also have my own preferences. Now it has more code and it's not super clear that it's returning a nil at the end (if you don't look up at the return values... but I thought that this is what we wanted to avoid).

Please, this have been like this for days if not weeks, next time, try to make this comments earlier.

Reply:
Thanks for adjusting it, appreciate it. Not sure i can follow your comment above as now if I look at the end of the method it's very clear, no error is returned.

Sorry that we missed that in previous reviews but I think initially we focused on the high level details like naming before going into the details.